include/eve/module/real/core/function/regular/simd/x86/compress_store.hpp

```diff
@@ -105,47 +178,132 @@ namespace eve::detail
     }
   }
 
-  template <typename T, typename N>
-  EVE_FORCEINLINE wide<T, N> permvar8(wide<T, N> v, __m256i pattern)
-    requires (current_api >= avx2)
+
+  /*
+    Credit again goes to @aqrit.
+
+    See comment in compress store for 8 elements first.
+    00 should become 0, 10|01 -> 1, 11 -> 2.
+    This is adding digigts.
+
+    We have 3 pairs, so the first 1 should correspond to power 0, second to 1, third 3 - 2
+
+    So we should & with [1, 1, 3, 3, 9, 9] and then reduce.
+
+    Now, the problem is there is not really a good instruction to reduce 8 numbers.
+    The closest we come is `_mm_sad_epu8`.
+    However, it sums absolute differences for bytes.
+    We could use a zero mask.
+    But instead we can use the same mask we got to mask the 1, 3, 9.
+    This get's us a somewhat inverted result:
+    0x0101 - 0x0101 becomes a 0 instead of 2.
+    So we can just take that into account when ordering masks.
+
+    Example:
+    (true, false, true, true, true, false, false, false)
+     1             3    3      9
+     reduced is 16
+     Total sum is 3^4 -1 = 26.
+     26 - 16 = 10
+     Resulting index is 10
+
+    Since we do reduction, we can also try to squese the popcount in it.
+    But, the popcount will also come inverted, so we need to subtract it from 8.
+    If native is suppored, might be not worth it again.
+
+    NOTE: @aqrit does not have the problem with an inverted mask nor subtracting from 8.
+          I don't understand how though so far.
+    NOTE: we could also use mmx intrinsics but tend to not use them.
+  */
+  template <real_scalar_value T, typename N>
+  EVE_FORCEINLINE
+  std::pair<std::uint32_t, std::uint16_t>
+  base_3_mask(eve::logical<eve::wide<T, N>> mask) {
+    if constexpr ( sizeof(T) == 4 && N() == 8 && current_api == avx2)
+    {
+      // Converting to byte per element
+      // This also messes with order a bit
+      // Alternative is to convert to epi16 using packus but this seems better.
+      std::int32_t to_4_bits_per_element = _mm256_movemask_epi8(mask);
+      std::uint32_t top_half             =  to_4_bits_per_element >> 4;
+      __m128i byte_mask = _mm_set_epi32(top_half, to_4_bits_per_element,
+                                        top_half, to_4_bits_per_element);
+
+      __m128i mask = _mm_set_epi32(
+        0x01010101, 0x01010101,  // popcount bytes
+        0x00090301, 0x00090301   // idx mask
+      );
+
+      __m128i reduced = _mm_sad_epu8(_mm_and_si128(byte_mask, mask), mask);
+      return {_mm_cvtsi128_si32(reduced), 8 - _mm_extract_epi16(reduced, 4)};
+    }
```

Comment:
Hmm

`_mm_packus_epi32 ` is actually 1.
But extract is expensive.

`_mm256_packus_epi32` seems also cheap.

And then I can try to use: `_mm256_sad_epu8`

Dunno, maybe it will be better, everything needs a measuring.
Also depends on how well I can do the 8 shorts.